sdk/core/System.ClientModel/src/Options/ClientLoggingOptions.cs

```diff
@@ -0,0 +1,277 @@
+            }
+            else
+            {
+                return _allowedQueryParameters ??= new ReadOnlyCollection<string>(DefaultAllowedQueryParameters);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Freeze this instance of <see cref="ClientLoggingOptions"/>.  After this method
+    /// has been called, any attempt to set properties on the instance or call
+    /// methods that would change its state will throw <see cref="InvalidOperationException"/>.
+    /// </summary>
+    public virtual void Freeze()
+    {
+        _frozen = true;
+        if (_allowedHeaderNames is not null)
+        {
+            ChangeTrackingStringList? headersChangeTracking = _allowedHeaderNames as ChangeTrackingStringList;
+
+            // Set the property to readonly before checking HasChanged
+            _allowedHeaderNames = new ReadOnlyCollection<string>(_allowedHeaderNames);
+            _allowedHeaderNamesHasChanged = headersChangeTracking?.HasChanged;
+        }
+        if (_allowedQueryParameters is not null)
+        {
+            ChangeTrackingStringList? queryParamsChangeTracking = _allowedQueryParameters as ChangeTrackingStringList;
+
+            // Set the property to readonly before checking HasChanged
+            _allowedQueryParameters = new ReadOnlyCollection<string>(_allowedQueryParameters);
+            _allowedQueryParametersHasChanged = queryParamsChangeTracking?.HasChanged;
+        }
+    }
+
+    /// <summary>
+    /// Assert that <see cref="Freeze"/> has not been called on this
+    /// <see cref="ClientLoggingOptions"/> instance.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">Thrown when an attempt is
+    /// made to change the state of this <see cref="ClientLoggingOptions"/> instance
+    /// after <see cref="Freeze"/> has been called.</exception>
+    protected void AssertNotFrozen()
+    {
+        if (_frozen)
+        {
+            throw new InvalidOperationException("Cannot change a MessageLoggingPolicyOptions instance after the ClientPipeline is created.");
+        }
+    }
+
+    internal void ValidateOptions()
+    {
+        if (EnableLogging == false
+            && (EnableMessageLogging == true || EnableMessageContentLogging == true))
+        {
+            throw new InvalidOperationException("HTTP Message logging cannot be enabled when client-wide logging is disabled.");
+        }
+        if (EnableMessageLogging == false
+            && EnableMessageContentLogging == true)
+        {
+            throw new InvalidOperationException("HTTP Message content logging cannot be enabled when HTTP message logging is disabled.");
+        }
+        if (EnableLogging == false && LoggerFactory != null) // TODO
+        {
+            throw new InvalidOperationException("An ILoggerFactory cannot be set if client-wide logging is disabled.");
+        }
+
+        // TODO - throw if content size is set but content logging is false?
```

Comment:
TODO - how far to go with validation...

Reply:
What about if logging is disabled but a message logging policy implementation is provided?

Reply:
We need to allow these to be specified regardless of whether or not logging is enabled.   There's a legitimate scenario where logging is controlled by configuration and may be off in local or staging environments but enabled in production.   

The expectation is that the same exact code gets promoted locally, then to staging, and ultimately to production with no changes.  Validation is performed in each environment and having to make code changes to add a logging implementation only in production where we want logging would be a blocker and policy violation.  This is prevalent in enterprise environments and particularly strict in industries such as Pharm and Finance.

Reply:
@jsquire thanks for that scenario, that's good to understand better. I wonder if this scenario is a reason to change the precedence rules I have currently set. With the current implementation, ClientPipelineOptions.ClientLoggingOptions.EnableLogging is the highest priority, the ClientPipeline.Create method will not add a logging policy if EnableLogging is false. E.g. this:
```
ClientLoggingOptions loggingOptions = new() { EnableLogging = false };
ClientPipelineOptions options = new()
{
    MessageLoggingPolicy = MyCustomPolicy(loggingOptions),
    ClientLoggingOptions = loggingOptions
};
```
will not add MyCustomPolicy to the pipeline. Although for the scenario you mentioned that should be fine.

There is a potential for confusion no matter what with the behavior is, so these exceptions were meant to help alleviate that, but I agree that no exception should be thrown in the scenario you said so I will not add that.